arroyo-sql/src/lib.rs

```diff
@@ -164,6 +175,83 @@ impl ArroyoSchemaProvider {
     pub fn add_defs(&mut self, source: impl Into<String>, defs: impl Into<String>) {
         self.source_defs.insert(source.into(), defs.into());
     }
+
+    fn get_inlined_source(
+        &self,
+        query: &Statement,
+    ) -> Result<(String, Vec<StructField>, Operator)> {
+        let Statement::CreateTable {
+        name,
+        columns,
+        with_options,
+        ..
+    } = query else {
+        bail!("expected create table statement");
+    };
+        let connection_name = with_options
+            .iter()
+            .filter_map(|option| match option.name.value.as_str() {
+                "connection" => Some(value_to_inner_string(&option.value)),
+                _ => None,
+            })
+            .next()
+            .ok_or_else(|| anyhow!("expect connection in with block"))??;
+        let connection = self
+            .connections
+            .get(&connection_name)
+            .ok_or_else(|| anyhow!("connection {} not found", connection_name))?;
+
+        let json_fields: Vec<StructField> = columns
+            .iter()
+            .map(|column| {
+                let name = column.name.value.to_string();
+                let data_type = convert_data_type(&column.data_type)?;
+                let nullable = !column
+                    .options
+                    .iter()
+                    .any(|option| matches!(option.option, ColumnOption::NotNull));
+                Ok(StructField {
+                    name,
+                    alias: None,
+                    data_type: TypeDef::DataType(data_type, nullable),
+                })
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        match connection.connection_type {
+            Some(ConnectionType::Kafka(ref kafka_connection)) => {
+                let topic: String = with_options
+                    .iter()
+                    .filter_map(|option| match option.name.value.as_str() {
+                        "topic" => Some(value_to_inner_string(&option.value)),
+                        _ => None,
+                    })
+                    .next()
+                    .ok_or_else(|| anyhow!("expect topic in with block"))??;
+                Ok((
+                    name.to_string(),
+                    json_fields,
+                    Operator::KafkaSource {
+                        topic,
+                        // split to a vec
+                        bootstrap_servers: kafka_connection
+                            .bootstrap_servers
+                            .split(",")
+                            .map(|s| s.to_string())
+                            .collect(),
+                        offset_mode: OffsetMode::Latest,
+                        kafka_input_format: SerializationMode::Json,
+                        // TODO: use setting for account
```

Comment:
this points to a bit of ugliness here, that we're having to solve this class of problems (configuring a kafka source) both here and in the API -- I don't have a good solution to it right now though

Reply:
Yeah, agree we need to figure out what to do. We could shove more relevant data into the ArroyoSchemaProvider, apply the constraints after coming out of SQL, or something else.